### ISSUE TITLE
Aligning the pointcloud to .obj .ply files created in Pix4D 

### ISSUE THREAD

**Issue by kayates** [potree/potree#968]
Disclaimer: Not a programmer, my greatest appreciation and respect for those of you who contribute to this cause.

I have pointcloud, mesh and texture files (,ply .mtl .jpg .obj) created, aligned and output from Pix4D.

After considerable effort through using the meshes.html example I have been able to successfully display the pointcloud, mesh and texture in the same window. I've hunted through lots of source code in the console to find coordinates via viewer.scene.pointclouds [PointCLoudOctree] and found coordinates that when input as is shown in the meshes.html example, get me really close... but its clear from the output that I have not found the accurate solution.

(A quick note for those struggling to be able to see the textures on their models... don't make the same mistake as me and rename the .mtl and .obj files without modifying the reference those two files make to one another. You'll find in the first few lines of each of the files that they refer to their original filename that they were assigned when created.)

At first I thought the solution would be as easy as using the coordinates found in the pcoGeometry offset found in the console as the origin for the mesh and texture files and replacing the same coordinates in my html file. Not so fast... it got me close and now I could at least see the file in the window, but it was clearly still not matching up. I made a number of attempts to match by using the information I could glean from the min and max bounding boxes, thinking that perhaps the origin of the pointcloud was at the center of the different bounding boxes. After trying all those combinations... still no luck. I've hunted through the internet to see if I can find a solution for a few days now and can't find one and if I am writing this then you knw I am desperate for a repeatable and simple solution.

I've concluded that I am either missing something so obvious that I have overcomplicated this issue or its well beyond my skillset in this area (see diclaimer at top). Either way, I am stuck and need your help. I would really appreciate a relatively simple way to be able to bring all these files together in a single view of potree. This will be an exercise I will need to repeat on multiple projects over time. Thank you in advance for your help.

Here is the code I have hacked together by copying and pasting from the meshes.html file to my own as well as a screen shot of the results I am getting. Thanks again for your help.

`	<script>
	
		window.viewer = new Potree.Viewer(document.getElementById("potree_render_area"));
		
		viewer.setFOV(60);
		viewer.setPointBudget(1*1000*10000);
		viewer.setMinNodeSize(0)
		viewer.useHQ = false;
		document.title = "";
		viewer.setEDLEnabled(false);
		viewer.setBackground("skybox"); // ["skybox", "gradient", "black", "white"];
		viewer.setDescription("weflyit3D");
		viewer.loadSettingsFromURL();
		viewer.setLengthUnit('ft');
		
		//viewer.setDescription("Point cloud courtesy of <a target='_blank' href='https://www.sigeom.ch/'>sigeom sa</a>");
		
		viewer.loadGUI(() => {
			viewer.setLanguage('en');
			$("#menu_appearance").next().show();
			$("#menu_tools").next().show();
			$("#menu_scene").next().show();
			$(".potree_menu_toggle").show();
			//$("#menu_appearance").hide();
			//$("#menu_tools").hide();
			//$("#menu_measurements").hide();
			//$("#menu_annotations").hide();
			//$("#menu_scene").hide();
			//$("#menu_classification").hide();
			//$("#menu_other_settings").hide();
			//$("#menu_about").hide();
			//viewer.toggleSidebar();
		});

		Potree.loadPointCloud("../pointclouds/house1/cloud.js", "house", function(e){
			viewer.scene.addPointCloud(e.pointcloud);
			
			let material = e.pointcloud.material;
			material.size = 1;
			material.pointSizeType = Potree.PointSizeType.ADAPTIVE;
			material.shape = Potree.PointShape.SQUARE;
			viewer.scene.view.pitch = -0.23800061030977313;
			viewer.scene.view.yaw = 0.42052578149868697;
			viewer.scene.view.position.set (1509621.0457389548, 7379810.423820436, 4606.0837179215105);
			viewer.scene.view.radius = 100;
			//viewer.fitToScreen();
		});		
		var loader = new THREE.PLYLoader();
		loader.load( "/potree-2.0/pages/models/house.ply", (geometry) => {
			geometry.computeVertexNormals();

			
		// place three instances of this bunny into the scene
			
			let mesh1;
			{
				let material = new THREE.MeshNormalMaterial();
				mesh1 = new THREE.Mesh( geometry, material );
				mesh1.position.set(1509573.77515138, 7379931.82550919, 4640.93999999999);
				mesh1.scale.multiplyScalar(1);
				viewer.scene.scene.add(mesh1);
			}
			viewer.onGUILoaded(() => {
			// Add entries to object list in sidebar
				let tree = $(`#jstree_scene`);
				let parentNode = "other";

				let houseID = tree.jstree('create_node', parentNode, { 
						"text": "House Mesh", 
						"icon": `${Potree.resourcePath}/icons/triangle.svg`,
						"data": mesh1
					}, 
					"last", false, false);
				tree.jstree(mesh1.visible ? "check_node" : "uncheck_node", houseID);
			});
			
		});

		{
			let manager = new THREE.LoadingManager();
			manager.onProgress = function ( item, loaded, total ) {
				console.log( item, loaded, total );
			};

			let onProgress = function ( xhr ) {
				if ( xhr.lengthComputable ) {
					let percentComplete = xhr.loaded / xhr.total * 100;
					console.log( Math.round(percentComplete, 2) + '% downloaded' );
				}
			};
			let onError = function ( xhr ) {};

			const texture = new THREE.TextureLoader().load("/potree-2.0/pages/models/house.jpg");

			texture.wrapS = THREE.RepeatWrapping;
			texture.wrapT = THREE.RepeatWrapping;

			let loader = new THREE.OBJLoader( manager );
			loader.load("/potree-2.0/pages/models/house.obj", function ( object ) {
				object.traverse( function ( child ) {
					if ( child instanceof THREE.Mesh ) {
						child.material.map = texture;
					}
				} );

				object.position.set(1509573.77515138, 7379931.82550919, 4640.93999999999);
				object.scale.multiplyScalar(1);				
				
				viewer.scene.scene.add( object );

				viewer.onGUILoaded(() => {
					// Add entries to object list in sidebar
					let tree = $(`#jstree_scene`);
					let parentNode = "other";

					let houseID = tree.jstree('create_node', parentNode, { 
							text: "House Textured", 
							icon: `${Potree.resourcePath}/icons/triangle.svg`,
							data: object
						}, 
						"last", false, false);
					tree.jstree(object.visible ? "check_node" : "uncheck_node", houseID);

					//tree.jstree("open_node", parentNode);
				});

			}, onProgress, onError );
		}

		{ // LIGHTS
			const directional = new THREE.DirectionalLight( 0xffffff, 1.0);
			directional.position.set( 10, 10, 10 );
			directional.lookAt(0, 0, 0);

			const ambient = new THREE.AmbientLight(0x555555);

			viewer.scene.scene.add(directional);
			viewer.scene.scene.add(ambient);
		}		
	</script> 
	
	
  </body>
</html>`

![image](https://user-images.githubusercontent.com/78123325/106079053-bcfc1300-60d1-11eb-99e2-5f4a1ba65fa2.png)
![image](https://user-images.githubusercontent.com/78123325/106079434-72c76180-60d2-11eb-9131-d6033dcbde62.png)
![image](https://user-images.githubusercontent.com/78123325/106079505-9b4f5b80-60d2-11eb-8ce3-83497061f24d.png)
![image](https://user-images.githubusercontent.com/78123325/106079588-c33ebf00-60d2-11eb-8990-9fccad919ed3.png)

**Comment by mikepianka** [comment]
I would check that your OBJ file is being exported with the same georeferenced location as the point cloud. Import your OBJ back into the software where your pointcloud is and confirm it lines up. Or open the OBJ in a text editor to look at the vertex coordinates. Normally you shouldn't have to set a position on the mesh when you bring it into potree.

**Comment by kayates** [comment]
Thank you for the reply. I've checked on that.... that unfortunately is a part of the problem. The origin of the point cloud seems to insert to a different location from the point cloud. Do you know how to find what origin is of the cloud is inserted at like the .obj and .ply?. They line up perfectly.

**Comment by mikepianka** [comment]
Look at the coordinates of each and check if it looks like one is in feet and the other meters. A multiple of 3.28. You can use the multiply scalar method on the obj to scale it if that's the case.

**Comment by kayates** [comment]
Thanks for the feedback. The suggestion to look at the units is one of the first things I looked at and I've verified that they are all the same. I believe if there were some way to know exactly the coordinates of where the cloud is placed in the viewer then I would think I could set the obj and ply files to that same coordinate. Note that the obj and the ply match up perfectly using a single coordinate. It seems there would be a similar point of insertion associated with the point cloud. I just dont' know what that variable/location is. That is what I am looking for... the insertion origin of the point cloud that corresponds with the same point in the obj and ply files. Thanks again for the help in advance.

**Comment by mikepianka** [comment]
If all of your data sources are on the same coordinate system you don't need to define an insertion point. The objects will be drawn based on their geometry's coordinates. You need to figure out purely based on the coordinates of each dataset whether they will align or not. Try bringing your point cloud, ply, and obj into some other software package to see if all 3 align. 

Also, you're doing some weird stuff with the scene yaw and pitch. Is that to set the default view of the scene? If so just set a scene lookAt instead. You can get the scene view position and lookAt coordinates from the camera in the toolbar layers.

**Comment by CallumMcK** [comment]
@kayates I am interested to see if you worked out your problem - I am having slightly similar issues!

**Comment by kayates** [comment]
Nope, never could get it to work with my resources. If you figure something out I would be very interested. Anyone?
